File: Objective-C Version/DownloadButtonDemo/CADemos/DownloadButton/DownloadButton.m
//
//  DownloadButton.m
//  CADemos
//
//  Created by Kitten Yang on 9/21/15.
//  Copyright (c) 2015 Kitten Yang. All rights reserved.
//

#import "DownloadButton.h"

@implementation DownloadButton{

    BOOL animating;
    CGRect originframe;
}

-(id)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        [self setUpSomething];
    }
    return self;
}

-(id)initWithCoder:(NSCoder *)aDecoder{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self setUpSomething];
    }
    return self;
}

-(id)init{
    self = [super init];
    if (self) {
        [self setUpSomething];
    }
    return self;
}

-(void)setUpSomething{
    UITapGestureRecognizer *tapGes = [[UITapGestureRecognizer alloc]initWithTarget:self action:@selector(tapped:)];
    [self addGestureRecognizer:tapGes];
}


#pragma mark -- UITapGesture

-(void)tapped:(UITapGestureRecognizer *)tapped{
    originframe = self.frame;
    
    if (animating == YES) {
        return;
    }
    
    for (CALayer *subLayer in self.layer.sublayers) {
        [subLayer removeFromSuperlayer];
    }
    
    self.backgroundColor = [UIColor colorWithRed:0.0 green:122/255.0 blue:255/255.0 alpha:1.0];
    
    animating = YES;
    
    //这个不是正确的方式，正确的方式是先改变model layer的属性，再应用动画。
//    radiusAnimation.fillMode = kCAFillModeForwards;
//    radiusAnimation.removedOnCompletion = NO;
    //WWDC 2011 视频 http://adcdownload.apple.com/videos/wwdc_2011__sd/session_421__core_animation_essentials.m4v
    self.layer.cornerRadius = self.progressBarHeight/2;
    
    //注意:[UIView animate]的方法里只能对UIView的属性进行动画，对于layer的属性是无效的。比如你在这里想让self.AnimateView.layer.cornerRadius = 50.0; 是没有意义的。必须使用CoreAnimation。
    CABasicAnimation *radiusShrinkAnimation = [CABasicAnimation animationWithKeyPath:@"cornerRadius"];
    radiusShrinkAnimation.duration = 0.2f;
    radiusShrinkAnimation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
    radiusShrinkAnimation.fromValue = @(originframe.size.height/2);
    
    //不需要设置toValue了
    radiusShrinkAnimation.delegate = self;
    [self.layer addAnimation:radiusShrinkAnimation forKey:@"cornerRadiusShrinkAnim"];

}


-(void)animationDidStart:(CAAnimation *)anim{
    
    //这里介绍两种方式区分不同的anim 1、对于加在一个全局变量上的anima，比如例子里的self.AnimateView ，这是一个全局变量，所以我们在这里可以通过[self.AnimateView.layer animationForKey:]根据动画不同的key来区分
    //2、然而对于一个非全局的变量，比如demo中的progressLayer，可以用KVO:[pathAnimation setValue:@"strokeEndAnimation" forKey:@"animationName"];注意这个animationName是我们自己设定的。
    
    if ([anim isEqual:[self.layer animationForKey:@"cornerRadiusShrinkAnim"]]) {
        
        [UIView animateWithDuration:0.6f delay:0.0f usingSpringWithDamping:0.6 initialSpringVelocity:0.0 options:UIViewAnimationOptionCurveEaseOut animations:^{
            self.bounds = CGRectMake(0, 0, _progressBarWidth, _progressBarHeight);
        } completion:^(BOOL finished) {
            [self.layer removeAllAnimations];
            [self progressBarAnimation];
        }];
        
    }else if ([anim isEqual:[self.layer animationForKey:@"cornerRadiusExpandAnim"]]){
        
        [UIView animateWithDuration:0.6f delay:0.0f usingSpringWithDamping:0.6 initialSpringVelocity:0.0 options:UIViewAnimationOptionCurveEaseOut animations:^{
            self.bounds = CGRectMake(0, 0, originframe.size.width, originframe.size.height);
            self.backgroundColor = [UIColor colorWithRed:0.1803921568627451 green:0.8 blue:0.44313725490196076 alpha:1.0];
        } completion:^(BOOL finished) {
            [self.layer removeAllAnimations];
            [self checkAnimation];
            //-----
            animating = NO;
        }];
        
    }
}


-(void)animationDidStop:(CAAnimation *)anim finished:(BOOL)flag{
    
    if ([[anim valueForKey:@"animationName"]isEqualToString:@"progressBarAnimation"]){
        
        [UIView animateWithDuration:0.3 animations:^{
            for (CALayer *subLayer in self.layer.sublayers) {
                subLayer.opacity = 0.0f;
            }
        } completion:^(BOOL finished) {
            if (finished) {
                for (CALayer *subLayer in self.layer.sublayers) {
                    [subLayer removeFromSuperlayer];
                }
                
                self.layer.cornerRadius = originframe.size.height/2;
                
                CABasicAnimation *radiusExpandAnimation = [CABasicAnimation animationWithKeyPath:@"cornerRadius"];
                radiusExpandAnimation.duration = 0.2f;
                radiusExpandAnimation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
                radiusExpandAnimation.fromValue = @(_progressBarHeight/2);
                
                radiusExpandAnimation.delegate = self;
                [self.layer addAnimation:radiusExpandAnimation forKey:@"cornerRadiusExpandAnim"];
                
            }
        }];
        
    }
}


#pragma mark -- Helper

-(void)progressBarAnimation{
    
    CAShapeLayer *progressLayer = [CAShapeLayer layer];
    UIBezierPath *path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(_progressBarHeight/2, self.bounds.size.height/2)];
    [path addLineToPoint:CGPointMake(self.bounds.size.width-_progressBarHeight/2, self.bounds.size.height/2)];
    
    progressLayer.path = path.CGPath;
    progressLayer.strokeColor = [UIColor whiteColor].CGColor;
    progressLayer.lineWidth = _progressBarHeight-6;
    progressLayer.lineCap = kCALineCapRound;
    //设置了kCALineCapRound 那么圆角弧度自动被设为 lineWidth/2 .所以要想进度条距离外围的间距相等，起始点的 x 坐标应该等于满足公式 x=lineWidth/2+space; ∵ lineWidth ＝ _progressBarHeight-space*2 ∴x = height/2.与 linewidth 是多少并没有关系
    [self.layer addSublayer:progressLayer];
    
    CABasicAnimation *pathAnimation = [CABasicAnimation animationWithKeyPath:@"strokeEnd"];
    pathAnimation.duration = 2.0f;
    pathAnimation.fromValue = @(0.0f);
    pathAnimation.toValue = @(1.0f);
    pathAnimation.delegate = self;
    [pathAnimation setValue:@"progressBarAnimation" forKey:@"animationName"];
    [progressLayer addAnimation:pathAnimation forKey:nil];
    
}

-(void)checkAnimation{
    
    CAShapeLayer *checkLayer = [CAShapeLayer layer];
    
    UIBezierPath *path = [UIBezierPath bezierPath];
    CGRect rectInCircle = CGRectInset(self.bounds, self.bounds.size.width*(1-1/sqrt(2.0))/2, self.bounds.size.width*(1-1/sqrt(2.0))/2);
    [path moveToPoint:CGPointMake(rectInCircle.origin.x + rectInCircle.size.width/9, rectInCircle.origin.y + rectInCircle.size.height*2/3)];
    [path addLineToPoint:CGPointMake(rectInCircle.origin.x + rectInCircle.size.width/3,rectInCircle.origin.y + rectInCircle.size.height*9/10)];
    [path addLineToPoint:CGPointMake(rectInCircle.origin.x + rectInCircle.size.width*8/10, rectInCircle.origin.y + rectInCircle.size.height*2/10)];
    
    checkLayer.path = path.CGPath;
    checkLayer.fillColor = [UIColor clearColor].CGColor;
    checkLayer.strokeColor = [UIColor whiteColor].CGColor;
    checkLayer.lineWidth = 10.0;
    checkLayer.lineCap = kCALineCapRound;
    checkLayer.lineJoin = kCALineJoinRound;
    [self.layer addSublayer:checkLayer];
    
    
    CABasicAnimation *checkAnimation = [CABasicAnimation animationWithKeyPath:@"strokeEnd"];
    checkAnimation.duration = 0.3f;
    checkAnimation.fromValue = @(0.0f);
    checkAnimation.toValue = @(1.0f);
    checkAnimation.delegate = self;
    [checkAnimation setValue:@"checkAnimation" forKey:@"animationName"];
    [checkLayer addAnimation:checkAnimation forKey:nil];

}

@end

File: Swift Version/CADemos-Swift/CADemos-Swift/Classes/DownloadButton.swift

//
//  DownloadButton.swift
//  CADemos-Swift
//
//  Created by Kitten Yang on 12/30/15.
//  Copyright © 2015 Kitten Yang. All rights reserved.
//

import UIKit

let kRadiusShrinkAnim = "cornerRadiusShrinkAnim"
let kRadiusExpandAnim = "radiusExpandAnimation"
let kProgressBarAnimation = "progressBarAnimation"
let kCheckAnimation = "checkAnimation"

class DownloadButton: UIView {

    var progressBarHeight: CGFloat = 0.0
    var progressBarWidth: CGFloat = 0.0
    private var originframe: CGRect = CGRectZero
    private var animating: Bool = false
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setUpViews()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setUpViews()
    }
    
    private func setUpViews(){
        let tapGes = UITapGestureRecognizer(target: self, action: "handleButtonDidTapped:")
        addGestureRecognizer(tapGes)
    }
    
    @objc private func handleButtonDidTapped(gesture: UITapGestureRecognizer){
        if animating {
            return
        }
        animating = true
        originframe = frame
        if let subLayers = layer.sublayers{
            for subLayer in subLayers{
                subLayer.removeFromSuperlayer()
            }
        }
        backgroundColor = UIColor(red: 0.0, green: 122/255.0, blue: 255/255.0, alpha: 1.0)
        layer.cornerRadius = progressBarHeight / 2
        let radiusShrinkAnimation = CABasicAnimation(keyPath: "cornerRadius")
        radiusShrinkAnimation.duration = 0.2
        radiusShrinkAnimation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
        radiusShrinkAnimation.fromValue = originframe.height / 2
        radiusShrinkAnimation.delegate = self
        layer.addAnimation(radiusShrinkAnimation, forKey: kRadiusShrinkAnim)
    }
    
    private func progressBarAnimation() {
        let progressLayer = CAShapeLayer()
        let path = UIBezierPath()
        path.moveToPoint(CGPoint(x: progressBarHeight / 2, y: self.frame.height / 2))
        path.addLineToPoint(CGPointMake(bounds.size.width - progressBarHeight / 2, bounds.size.height / 2))
        
        progressLayer.path = path.CGPath
        progressLayer.strokeColor = UIColor.whiteColor().CGColor
        progressLayer.lineWidth = progressBarHeight - 6
        progressLayer.lineCap = kCALineCapRound
        layer.addSublayer(progressLayer)
        
        let pathAnimation = CABasicAnimation(keyPath: "strokeEnd")
        pathAnimation.duration = 2.0
        pathAnimation.fromValue = 0.0
        pathAnimation.toValue = 1.0
        pathAnimation.delegate = self
        pathAnimation.setValue(kProgressBarAnimation, forKey: "animationName")
        progressLayer.addAnimation(pathAnimation, forKey: nil)
    }
    
    private func checkAnimation() {
        let checkLayer = CAShapeLayer()
        let path = UIBezierPath()
        let rectInCircle = CGRectInset(self.bounds, self.bounds.size.width*(1-1/sqrt(2.0))/2, self.bounds.size.width*(1-1/sqrt(2.0))/2)
        path.moveToPoint(CGPoint(x: rectInCircle.origin.x + rectInCircle.size.width/9, y: rectInCircle.origin.y + rectInCircle.size.height*2/3))
        path.addLineToPoint(CGPoint(x: rectInCircle.origin.x + rectInCircle.size.width/3,y: rectInCircle.origin.y + rectInCircle.size.height*9/10))
        path.addLineToPoint(CGPoint(x: rectInCircle.origin.x + rectInCircle.size.width*8/10, y: rectInCircle.origin.y + rectInCircle.size.height*2/10))
        
        checkLayer.path = path.CGPath
        checkLayer.fillColor = UIColor.clearColor().CGColor
        checkLayer.strokeColor = UIColor.whiteColor().CGColor
        checkLayer.lineWidth = 10.0
        checkLayer.lineCap = kCALineCapRound
        checkLayer.lineJoin = kCALineJoinRound
        self.layer.addSublayer(checkLayer)

        let checkAnimation = CABasicAnimation(keyPath: "strokeEnd")
        checkAnimation.duration = 0.3;
        checkAnimation.fromValue = 0.0
        checkAnimation.toValue = 1.0
        checkAnimation.delegate = self
        checkAnimation.setValue(kCheckAnimation, forKey:"animationName")
        checkLayer.addAnimation(checkAnimation,forKey:nil)
    }
}

extension DownloadButton {
    
    // MARK : CAAnimationDelegate
    
    override func animationDidStart(anim: CAAnimation) {
        if anim.isEqual(self.layer.animationForKey(kRadiusShrinkAnim)) {
            UIView.animateWithDuration(0.6, delay: 0.0, usingSpringWithDamping: 0.6, initialSpringVelocity: 0.0, options: .CurveEaseOut, animations: { () -> Void in
                self.bounds = CGRect(x: 0, y: 0, width: self.progressBarWidth, height: self.progressBarHeight)
                }, completion: { (finished) -> Void in
                    self.layer.removeAllAnimations()
                    self.progressBarAnimation()
            })
        }else if anim.isEqual(self.layer.animationForKey(kRadiusExpandAnim)){
            UIView.animateWithDuration(0.6, delay: 0.0, usingSpringWithDamping: 0.6, initialSpringVelocity: 0.0, options: .CurveEaseOut, animations: { () -> Void in
                self.bounds = CGRect(x: 0, y: 0, width: self.originframe.width, height: self.originframe.height)
                self.backgroundColor = UIColor(red: 0.1803921568627451, green: 0.8, blue: 0.44313725490196076, alpha: 1.0)
                }, completion: { (finished) -> Void in
                    self.layer.removeAllAnimations()
                    self.checkAnimation()
                    self.animating = false
            })
        }
        
    }
    
    override func animationDidStop(anim: CAAnimation, finished flag: Bool) {
        if let animationName = anim.valueForKey("animationName") where
            animationName.isEqualToString(kProgressBarAnimation) {
                UIView.animateWithDuration(0.3, animations: { () -> Void in
                    if let sublayers = self.layer.sublayers{
                        for subLayer in sublayers {
                            subLayer.opacity = 0.0
                        }
                    }
                    }, completion: { (finished) -> Void in
                        if let sublayers = self.layer.sublayers{
                            for sublayer in sublayers {
                                sublayer.removeFromSuperlayer()
                            }
                        }
                        self.layer.cornerRadius = self.originframe.height / 2
                        let radiusExpandAnimation = CABasicAnimation(keyPath: "cornerRadius")
                        radiusExpandAnimation.duration = 0.2
                        radiusExpandAnimation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
                        radiusExpandAnimation.fromValue = self.progressBarHeight / 2
                        radiusExpandAnimation.delegate = self
                        self.layer.addAnimation(radiusExpandAnimation, forKey: kRadiusExpandAnim)
                })
        }
    }
}

